client/src/components/WorkflowInvocationState/InvocationStepsProgressBar.vue

```diff
@@ -0,0 +1,68 @@
+<script setup lang="ts">
+import { computed } from "vue";
+
+import { InvocationStep, WorkflowInvocationElementView } from "@/api/invocations";
+import ProgressBar from "@/components/ProgressBar.vue";
+
+interface Props {
+    invocation?: WorkflowInvocationElementView;
+    invocationState: string;
+    invocationSchedulingTerminal: boolean;
+}
+
+const props = defineProps<Props>();
+
+const stepCount = computed<number>(() => {
+    return props.invocation?.steps.length || 0;
+});
+
+type StepStateType = { [state: string]: number };
+
+const stepStates = computed<StepStateType>(() => {
+    const stepStates: StepStateType = {};
+    if (!props.invocation) {
+        return {};
+    }
+    const steps: InvocationStep[] = props.invocation?.steps || [];
+    for (const step of steps) {
+        if (!step) {
+            continue;
+        }
+        // the API defined state here allowing null and undefined is odd...
```

Comment:
That's under your control, the table in https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields is helpful.

If the API always returns a null field and never just omits the value (which I guess we never do) you can remove the default value in https://github.com/galaxyproject/galaxy/blob/74766a3fb591839dfd5b15d8476a8d0eaf04423a/lib/galaxy/schema/invocation.py#L375.

Reply:
I hate the rabbit holes but it wasn't uninteresting... https://github.com/galaxyproject/galaxy/pull/17660/commits/3c030a8bb227d7916295190124dccccc1c08b057. I've revised the models so setting the legacy_job_view parameter returns a separate model with the older eccentricities hidden there.